plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -366,20 +381,46 @@ public ColumnMetadata getColumnMetadata(
     public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
     {
         log.debug("listTableColumns(session=%s, prefix=%s)", session, prefix);
-        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
         List<SchemaTableName> tables = prefix.toOptionalSchemaTableName()
                 .<List<SchemaTableName>>map(ImmutableList::of)
                 .orElseGet(() -> listTables(session, prefix.getSchema()));
-        for (SchemaTableName tableName : tables) {
-            try {
-                Optional.ofNullable(getTableHandleIgnoringConflicts(session, tableName))
-                        .ifPresent(tableHandle -> columns.put(tableName, getTableMetadata(session, tableHandle).getColumns()));
-            }
-            catch (TableNotFoundException e) {
-                // table disappeared during listing operation
-            }
+
+        List<BigQueryTableHandle> tableHandles = processInParallel(tables, table -> getTableHandleIgnoringConflicts(session, table))
+                .map(BigQueryTableHandle.class::cast)
+                .collect(toImmutableList());
+
+        return processInParallel(tableHandles, tableHandle -> safeGetTableMetadata(session, tableHandle))
+                .collect(toImmutableMap(ConnectorTableMetadata::getTable, ConnectorTableMetadata::getColumns));
+    }
+
+    @Nullable
+    private ConnectorTableMetadata safeGetTableMetadata(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        try {
+            return getTableMetadata(session, tableHandle);
+        }
+        catch (TableNotFoundException e) {
+            // table disappeared during listing operation
+            return null;
+        }
+    }
+
+    protected <T, R> Stream<R> processInParallel(List<T> list, Function<T, R> function)
+    {
+        List<ListenableFuture<R>> futures = list.stream()
+                .map(element -> executorService.submit(() -> function.apply(element)))
+                .collect(toImmutableList());
+        try {
+            return allAsList(futures).get().stream()
+                    .filter(Objects::nonNull);
```

Comment:
I think filtering should happen in a different level. It is part of semantic of why this method was called. `processInParallel` should be more generic.